plugin/storage/badger/spanstore/reader.go

```diff
@@ -207,12 +207,17 @@ func (r *TraceReader) scanTimeRange(plan *executionPlan) ([]model.TraceID, error
 	})
 
 	sort.Slice(traceKeys, func(k, h int) bool {
+		// This sorts by timestamp to descending order
 		return bytes.Compare(traceKeys[k][sizeOfTraceID+1:sizeOfTraceID+1+8], traceKeys[h][sizeOfTraceID+1:sizeOfTraceID+1+8]) > 0
 	})
 
-	traceIDs := make([]model.TraceID, len(traceKeys))
+	sizeCount := len(traceKeys)
+	if plan.limit > 0 {
+		sizeCount = plan.limit
+	}
+	traceIDs := make([]model.TraceID, sizeCount)
 
-	for i := 0; i < len(traceKeys); i++ {
+	for i := 0; i < len(traceKeys) && i < sizeCount; i++ {
```

Comment:
Only needs to be `i < sizeCount`

Reply:
Actually that alone would cause error (indexoutofbounds), due to the previous lines where sizeCount could be bigger than len(traceKeys).